Docs day acl improvements (#12460)
hashicorp/consul: website/content/docs/security/acl/acl-roles.mdx

---
layout: docs
page_title: Roles
description: >-
  This topic describes roles within the access control list (ACL) system. A role is a named set of policies and service identities.

>   They are intended to

Author: ```suggestion
  They enable you to reuse policies by decoupling the policies from the token distributed to team members.
```

```shell-session
$ curl --request PUT --data @payload.json http://127.0.0.1:8500/v1/acl/role
```

Refer to the [API documentation](/api-docs/acl/roles) for details.

## Role Attributes

Roles may contain the following table describe the attributes:

- `ID`: The `ID` is an auto-generated public identifier. You can specify the role `ID` when linking it to tokens.
- `Name`: A unique meaningful name for the role. You can specify the role `Name` when linking it to tokens.
- `Description`: (Optional) A human-readable description of the role.
- `Policies`: Specifies a the list of policies that are applicable for the role. The object can reference the policy `ID` or `Name` attribute.
- `ServiceIdentities`: Specifies a list of services that are applicable for the role. See [Service Identities](#service-identities) for details.
- `NodeIdentities`: Specifies a list of nodes that are applicable for the role. See [Node Identities](#node-identities) for details.

> - `Namespace`: <EnterpriseAlert inline /> The namespace this policy resides within. Roles can only be linked to policies that are defined in the same namespace and admin partition. See [Namespaces](/docs/enterprise/namespaces) and [Admin Partitions](/docs/enterprise/admin-partitions) for additional information. Requires Consul Enterprise 1.7.0+.

Author: ```suggestion
- `Namespace`: <EnterpriseAlert inline /> The namespace that the policy resides in. Roles can only be linked to policies that are defined in the same namespace. See [Namespaces](/docs/enterprise/namespaces) for additional information. Requires Consul Enterprise 1.7.0+
- `Partition`: <EnterpriseAlert inline/> The admin partition that the policy resides in. Roles can only be linked to policies that are defined in the same admin partition. See [Admin Partitions](/docs/enterprise/admin-partitions) for additional information. Requires Consul Enterprise 1.10.0+.
```

## Service Identities

> <!-- -> Added in Consul 1.5.0 # Remove and lean on versioning?-->

Author: Sort of -- I left it in there and commented out in case anyone disagreed with removing the kinds of callouts.

Node identities enable you to quickly construct policies for nodes, rather than manually creating identical polices for each node. They are used during the authorization process to automatically generate a policy for the node(s) specifed. You can specify the token linked to the policy in the [`acl_tokens_agent`](/docs/agent/options#acl_tokens_agent) field when configuring the agent. 

### Node Identity Specification

Use the following syntax to define a node identity:

<CodeTabs>

```json
{
  "NodeIdentities": [
    {
      "NodeName": "<node name>",
      "Datacenters": ["<datacenter name>"]
    }
  ]
}
```

```hcl
NodeIdentities = {
  NodeName = "<node name>"
  Datacenters = ["<datacenter name>"]
}
```

</CodeTabs>

- `NodeIdentities`: Declares a node identity block.
- `NodeIdentities.ServiceName`: String value that specifies the name of the node you want to associate with the policy.
- `NodeIdentitites.Datacenters`: Array that specifies the names of datacenters in which the node identity applies. This field is optional.

Refer to the the [API documentation for roles](/api/acl/roles#sample-payload) for additional information and examples.

> -> **Consul Enterprise Namespacing** - Node Identities can only be applied to tokens and roles in the `default` namespace. The generated policy rules allow for `service:read` permissions on all services in all namespaces.

Author: I understood that as a limitation. I'll follow up with someone after pushing these changes

> -> **Consul Enterprise Namespacing** - Node Identities can only be applied to tokens and roles in the `default` namespace. The generated policy rules allow for `service:read` permissions on all services in all namespaces.

Reviewer: This is probably because nodes are scoped to admin partitions, not to namespaces. For that reason, it wouldn't make sense to be able to specify a node identity in a namespace... except the special `default` namespace which is a catch-all for things not added to a user-created namespace (since every token must has a namespace specified).

How to communicate this in a way that's clear to the reader is a different question. I agree that a reader may ask the question @karl-cardenas-coding posed.

## Service Identities

<!-- -> Added in Consul 1.5.0 # Remove and lean on versioning?-->

You can specify a service identity when configuring roles or linking tokens to policies. Service identities are used during the authorization process to automatically generate a policy for the service(s) specifed. The policy will be linked to the role or token so that the service(s) can _be discovered_ and _discover other healthy service instances_ in a service mesh. See [Service Mesh](/docs/connect) for additional information about Consul service mesh. Service identities enable you to quickly construct policies for services, rather than creating identical polices for each service.

### Service Identity Specification

Use the following syntax to define a service identity:

<CodeTabs>
<CodeBlockConfig>

```json
{
  "ServiceIdentities": [
    {
      "ServiceName": "<service name>",
      "Datacenters": ["<datacenter name>"]
    }
  ]
}
```

</CodeBlockConfig>
<CodeBlockConfig>

```hcl
"ServiceIdentities" = {
  "ServiceName" = "<service name>"
  "Datacenters" = ["<datacenter name>"]
}
```

> </CodeBlockConfig>

Reviewer 2: ```suggestion
```

Per the `ServiceIdentitites.Datacenters` configuration, the `db` policy is scoped to resources in the `dc1` datacenter.

<CodeBlockConfig filename="db-policy.hcl">

```hcl
# Allow the service and its sidecar proxy to register into the catalog.
service "db" {
	policy = "write"
}
service "db-sidecar-proxy" {
	policy = "write"
}

# Allow for any potential upstreams to be resolved.
service_prefix "" {
    policy = "read"
}
node_prefix "" {
    policy = "read"
}
```

</CodeBlockConfig>

## Node Identities

<!-- -> Added in Consul 1.8.1 -- remove and lean on doc version? -->

You can specify a node identity when configuring roles or linking tokens to policies. Node identities are used during the authorization process to automatically generate a policy for the node(s) specifed. In most cases, "node" refers to a Consul agent.

You can specify the token linked to the policy in the [`acl_tokens_agent`](/docs/agent/options#acl_tokens_agent) field when configuring the agent. Node identities enable you to quickly construct policies for nodes, rather than creating identical polices for each node.

### Node Identity Specification

Use the following syntax to define a node identity:

<CodeTabs>

> <CodeBlockConfig>

Reviewer 2: ```suggestion
```

-> **Consul Enterprise Namespacing** - Node Identities can only be applied to tokens and roles in the `default` namespace. The generated policy rules allow for `service:read` permissions on all services in all namespaces.

The following policy is generated for each node when a node identity is declared:

```hcl
# Allow the agent to register its own node in the Catalog and update its network coordinates
node "<node name>" {
  policy = "write"
}

# Allows the agent to detect and diff services registered to itself. This is used during
# anti-entropy to reconcile difference between the agents knowledge of registered
# services and checks in comparison with what is known in the Catalog.
service_prefix "" {
  policy = "read"
}
```

Refer to the [rules reference](/docs/security/acl/acl-rules) for information about the rules in the policy.

### Example

The following role configuration contains a node identity for `node-1`. Note that the node identity is also scoped to the `dc2` datacenter. As a result, the policy will only be applied to nodes named `node-1` in `dc2`.

<CodeTabs>
<CodeBlockConfig filename="example-role.hcl">

```hcl

> "Description" = "Showcases all input parameters"
> "Name" = "example-role"
> "NodeIdentities" = {
>   "Datacenter" = "dc2"
>   "NodeName" = "node-1",
> }
> "Policies" = {
>   "ID" = "783beef3-783f-f41f-7422-7087dc272765"
> }
> "Policies" = {
>   "Name" = "node-read"
> }

Reviewer 2: ```suggestion
Description = "Showcases all input parameters"
Name = "example-role"
NodeIdentities = {
  Datacenter = "dc2"
  NodeName = "node-1",
}
Policies = {
  ID = "783beef3-783f-f41f-7422-7087dc272765"
}
Policies = {
  Name = "node-read"
}
```
